compiler/src/iree/compiler/Codegen/Common/CombineLayoutTransformation.cpp

```diff
@@ -0,0 +1,383 @@
+    Value boundValue = getValueOrCreateConstantIndexOp(rewriter, loc, bound);
+    auto isOutOfBounds =
+        rewriter
+            .create<arith::CmpIOp>(loc, arith::CmpIPredicate::uge, srcIdx,
+                                   boundValue)
+            ->getResult(0);
+    mask = rewriter.create<arith::AndIOp>(loc, mask, isOutOfBounds);
+  }
+  rewriter.modifyOpInPlace(yieldOp, [&]() {
+    yieldOp->setOperand(yieldOp->getNumOperands() - 1, mask);
+  });
+  rewriter.modifyOpInPlace(mapScatterOp, [&]() {
+    mapScatterOp->setOperand(0, extractSliceOp.getSource());
+  });
+  return mapScatterOp;
+}
+
+/// Fold the `op` into the `mapScatterOp`, if possible. The resulting
+/// map_scatter op is returned, if the `op` was folded. Otherwise, return
+/// failure.
+static FailureOr<IREE::LinalgExt::MapScatterOp>
+foldIntoMapScatter(RewriterBase &rewriter, Operation *op,
+                   IREE::LinalgExt::MapScatterOp mapScatterOp) {
+  if (isa<linalg::CopyOp>(op)) {
+    return foldIdentityLikeOpIntoMapScatter(rewriter, op, mapScatterOp);
+  }
+  if (auto transposeOp = dyn_cast<linalg::TransposeOp>(op)) {
+    return foldTransposeIntoMapScatter(rewriter, transposeOp, mapScatterOp);
+  }
+  if (auto expandShapeOp = dyn_cast<tensor::ExpandShapeOp>(op)) {
+    return foldExpandShapeIntoMapScatter(rewriter, expandShapeOp, mapScatterOp);
+  }
+  if (auto collapseShapeOp = dyn_cast<tensor::CollapseShapeOp>(op)) {
+    return foldCollapseShapeIntoMapScatter(rewriter, collapseShapeOp,
+                                           mapScatterOp);
+  }
+  if (auto extractSliceOp = dyn_cast<tensor::ExtractSliceOp>(op)) {
+    return foldExtractSliceIntoMapScatter(rewriter, extractSliceOp,
+                                          mapScatterOp);
+  }
+  return failure();
+}
+
+/// Return true if the mapScatterOp is an identity transformation, meaning it
+/// is just doing a copy from input to output. This is true if the mask always
+/// evaluates to true, and the mapping from input to output index is identity.
+static bool isIdentityMapScatterOp(IREE::LinalgExt::MapScatterOp mapScatterOp) {
+  if (mapScatterOp.getInputType() != mapScatterOp.getOutputType()) {
+    return false;
+  }
+
+  // Check that the mask is always true.
+  Block &transformBody =
+      mapScatterOp.getTransformationRegion().getBlocks().front();
+  auto yieldOp = cast<IREE::LinalgExt::YieldOp>(transformBody.getTerminator());
+  Value mask = yieldOp->getOperands().back();
+  std::optional<int64_t> constMask = getConstantIntValue(mask);
+  if (!constMask.has_value() || constMask.value() == 0) {
+    return false;
+  }
+
+  // Check that the block arguments are directly yielded in the order that they
+  // are defined in the block.
+  for (int i = 0; i < mapScatterOp.getOutputRank(); ++i) {
+    auto yieldedBbArg = dyn_cast<BlockArgument>(yieldOp.getOperand(i));
+    if (!yieldedBbArg || yieldedBbArg.getArgNumber() != i) {
+      return false;
+    }
+  }
```

Comment:
This is wrong because you could yield external block argument. It is a valid IR. You should check if the values match the block arguments of `transformBody`.

Counter example:

```mlir
func.func @map_scatter_tensor_static(
    %input: tensor<4x16x64xf32>, %output: tensor<4x16x64xf32>, %external_idx : index
) -> tensor<4x16x64xf32> {
  %0 = iree_linalg_ext.map_scatter %input into %output {
    ^bb0(%idx0: index, %idx1: index, %idx2: index):
      %mask = arith.constant true
      iree_linalg_ext.yield %idx0, %idx1, %external_idx, %mask : index, index, index, i1
  } : tensor<4x16x64xf32> into tensor<4x16x64xf32> -> tensor<4x16x64xf32>
  return %0 : tensor<4x16x64xf32>
}
```